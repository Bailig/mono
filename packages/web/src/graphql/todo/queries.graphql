query getTodos {
  todos {
    id
    content
    isCompleted
  }
}
