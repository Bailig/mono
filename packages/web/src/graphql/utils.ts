import { DataProxy } from "apollo-cache";
import { DocumentNode } from "graphql";

export const updateQueryHelper = <T>(
  store: DataProxy,
  query: DocumentNode,
  updateQuery: (queryData: T) => T,
): void => {
  const queryData = store.readQuery<T>({ query });
  if (!queryData) throw new Error("Queried data does not exist in cache");

  const newData = updateQuery(queryData);
  store.writeQuery({
    query,
    data: newData,
  });
};
