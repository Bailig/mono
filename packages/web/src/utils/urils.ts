import * as R from "ramda";

export const findIndexById: FindIndexById = (id, items) =>
  R.findIndex(R.propEq("id", id), items);

type FindIndexById = (id: string, items: { id: string }[]) => number;
