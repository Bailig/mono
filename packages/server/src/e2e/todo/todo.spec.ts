import { CallServer, startTestServer, toPromise } from "../utils";
import {
  CompleteTodoDocument,
  CreateTodoDocument,
  DeleteTodoDocument,
  UpdateTodoContentDocument,
} from "./mutations";
import { GetTodosDocument } from "./queries";

describe("todo", () => {
  let stop: () => void;
  let callServer: CallServer;
  const todo = { content: "test todo" };
  let id: string;

  beforeAll(async () => {
    const testServer = await startTestServer();
    stop = testServer.stop;
    callServer = testServer.callServer;
  });

  afterAll(() => stop && stop());

  it("should be created", async () => {
    const res = await toPromise(
      callServer({
        query: CreateTodoDocument,
        variables: { input: todo },
      }),
    );
    id = res.data?.createTodo.todo.id;

    expect(res).toMatchSnapshot({
      data: {
        createTodo: {
          todo: { id: expect.any(String) },
        },
      },
    });
  });

  it("shouldn't be created with empty content", async () => {
    const res = await toPromise(
      callServer({
        query: CreateTodoDocument,
        variables: { input: { content: "  " } },
      }),
    );

    expect(res).toMatchSnapshot();
  });

  it("should be queried as an array", async () => {
    const res = await toPromise(callServer({ query: GetTodosDocument }));

    expect(res).toMatchSnapshot({
      data: {
        todos: [{ id: expect.any(String) }],
      },
    });
  });

  it("should be updated with new content", async () => {
    const res = await toPromise(
      callServer({
        query: UpdateTodoContentDocument,
        variables: { input: { id, content: "test todo updated" } },
      }),
    );

    expect(res).toMatchSnapshot({
      data: {
        updateTodoContent: {
          todo: { id: expect.any(String) },
        },
      },
    });
  });

  it("shouldn't be updated with empty content", async () => {
    const res = await toPromise(
      callServer({
        query: UpdateTodoContentDocument,
        variables: { input: { id, content: " " } },
      }),
    );

    expect(res).toMatchSnapshot();
  });

  it("should be completed", async () => {
    const res = await toPromise(
      callServer({
        query: CompleteTodoDocument,
        variables: { input: { id } },
      }),
    );

    expect(res).toMatchSnapshot({
      data: {
        completeTodo: {
          todo: { id: expect.any(String) },
        },
      },
    });
  });

  it("should be deleted", async () => {
    const res = await toPromise(
      callServer({
        query: DeleteTodoDocument,
        variables: { input: { id } },
      }),
    );

    expect(res).toMatchSnapshot({
      data: {
        deleteTodo: {
          deletedTodoId: expect.any(String),
        },
      },
    });
  });
});
