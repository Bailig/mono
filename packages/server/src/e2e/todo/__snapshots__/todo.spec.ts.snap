// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo should be completed 1`] = `
Object {
  "data": Object {
    "completeTodo": Object {
      "todo": Object {
        "content": "test todo updated",
        "id": Any<String>,
        "isCompleted": true,
      },
      "userErrors": Array [],
    },
  },
}
`;

exports[`todo should be created 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "todo": Object {
        "content": "test todo",
        "id": Any<String>,
        "isCompleted": false,
      },
      "userErrors": Array [],
    },
  },
}
`;

exports[`todo should be deleted 1`] = `
Object {
  "data": Object {
    "deleteTodo": Object {
      "deletedTodoId": Any<String>,
      "userErrors": Array [],
    },
  },
}
`;

exports[`todo should be queried as an array 1`] = `
Object {
  "data": Object {
    "todos": Array [
      Object {
        "content": "test todo",
        "id": Any<String>,
        "isCompleted": false,
      },
    ],
  },
}
`;

exports[`todo should be updated with new content 1`] = `
Object {
  "data": Object {
    "updateTodoContent": Object {
      "todo": Object {
        "content": "test todo updated",
        "id": Any<String>,
        "isCompleted": false,
      },
      "userErrors": Array [],
    },
  },
}
`;

exports[`todo shouldn't be created with empty content 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "todo": null,
      "userErrors": Array [
        Object {
          "message": "content must be at least 1 characters",
          "path": "content",
        },
      ],
    },
  },
}
`;

exports[`todo shouldn't be updated with empty content 1`] = `
Object {
  "data": Object {
    "updateTodoContent": Object {
      "todo": null,
      "userErrors": Array [
        Object {
          "message": "content must be at least 1 characters",
          "path": "content",
        },
      ],
    },
  },
}
`;
